meteor/server/publications/partsUI/publication.ts

```diff
@@ -0,0 +1,277 @@
+
+	invalidateRundownIds: RundownId[]
+	invalidateSegmentIds: SegmentId[]
+	invalidatePartIds: PartId[]
+
+	invalidateQuickLoop: boolean
+}
+
+type RundownPlaylistFields = '_id' | 'studioId' | 'rundownIdsInOrder'
+const rundownPlaylistFieldSpecifier = literal<
+	MongoFieldSpecifierOnesStrict<Pick<DBRundownPlaylist, RundownPlaylistFields>>
+>({
+	_id: 1,
+	studioId: 1,
+	rundownIdsInOrder: 1,
+})
+
+async function setupUIPartsPublicationObservers(
+	args: ReadonlyDeep<UIPartsArgs>,
+	triggerUpdate: TriggerUpdate<UIPartsUpdateProps>
+): Promise<LiveQueryHandle[]> {
+	const playlist = (await RundownPlaylists.findOneAsync(args.playlistId, {
+		projection: rundownPlaylistFieldSpecifier,
+	})) as Pick<DBRundownPlaylist, RundownPlaylistFields> | undefined
+	if (!playlist) throw new Error(`RundownPlaylist "${args.playlistId}" not found!`)
+
+	const rundownsObserver = new RundownsObserver(playlist.studioId, playlist._id, (rundownIds) => {
+		logger.silly(`Creating new RundownContentObserver`)
+
+		const cache = createReactiveContentCache()
+
+		// Push update
+		triggerUpdate({ newCache: cache })
+
+		const obs1 = new RundownContentObserver(playlist.studioId, playlist._id, rundownIds, cache)
+
+		const innerQueries = [
+			cache.Segments.find({}).observeChanges({
+				added: (id) => triggerUpdate({ invalidateSegmentIds: [protectString(id)] }),
+				changed: (id) => triggerUpdate({ invalidateSegmentIds: [protectString(id)] }),
+				removed: (id) => triggerUpdate({ invalidateSegmentIds: [protectString(id)] }),
+			}),
+			cache.Parts.find({}).observeChanges({
+				added: (id) => triggerUpdate({ invalidatePartIds: [protectString(id)] }),
+				changed: (id) => triggerUpdate({ invalidatePartIds: [protectString(id)] }),
+				removed: (id) => triggerUpdate({ invalidatePartIds: [protectString(id)] }),
+			}),
+			cache.RundownPlaylists.find({}).observeChanges({
+				added: () => triggerUpdate({ invalidateQuickLoop: true }),
+				changed: () => triggerUpdate({ invalidateQuickLoop: true }),
+				removed: () => triggerUpdate({ invalidateQuickLoop: true }),
+			}),
+			cache.Studios.find({}).observeChanges({
+				added: () => triggerUpdate({ invalidateQuickLoop: true }),
+				changed: () => triggerUpdate({ invalidateQuickLoop: true }),
+				removed: () => triggerUpdate({ invalidateQuickLoop: true }),
+			}),
+		]
+
+		return () => {
+			obs1.dispose()
+
+			for (const query of innerQueries) {
+				query.stop()
+			}
+		}
+	})
+
+	// Set up observers:
+	return [rundownsObserver]
+}
+
+export async function manipulateUIPartsPublicationData(
+	_args: UIPartsArgs,
+	state: Partial<UIPartsState>,
+	collection: CustomPublishCollection<DBPart>,
+	updateProps: Partial<ReadonlyDeep<UIPartsUpdateProps>> | undefined
+): Promise<void> {
+	// Prepare data for publication:
+
+	// We know that `collection` does diffing when 'commiting' all of the changes we have made
+	// meaning that for anything we will call `replace()` on, we can `remove()` it first for no extra cost
```

Comment:
I think the docs of that function are misleading/incorrect? It is calculating a shallow difference, but it is comparing each property with a deep equality check.
But even so, inside the `replace` call, it is doing a clone of whatever you pass in, so its never going to be a cheap/quick diff.

So yes, I think this does need to be cleverer on what it updates each time it is run. 
At least handling the cases which can be easily cheap, such as when a part/segment changes it shouldnt need to invalidate everything. When the playlist quickLoop prop changes, I expect this could get very complicated to figure out exactly what should be invalidated. But it also won't change too often, so could get away with being less optimal.

Or perhaps take inspiration from what I did in the PieceContentStatus publication. Another approach would be to re-generate each Part as it changes. Then each time the publication changes, iterate through each part in the collection and update just the properties which quickloop may override. Be aware that the `updateAll` and `updateOne` methods of the collection don't perform any diffing themselves (and will always perform a clone), so manually diffing of the changes will be required to avoid making the ddp layer doing a deep comparison